tests/Aws/Tests/S3/StreamWrapperTest.php

```diff
@@ -591,6 +594,76 @@ public function testStatsRegularObjects()
         $this->assertEquals($ts, $stat['ctime']);
     }
 
+    /**
+     * Test that a cache can be attached and detached.
+     */
+    public function testAttachCache() {
+        StreamWrapper::attachCache(new DoctrineCacheAdapter(new ArrayCache()));
+        $this->assertTrue(StreamWrapper::hasCache());
+        StreamWrapper::detachCache();
+        $this->assertFalse(StreamWrapper::hasCache());
+    }
+
+    /**
+     * Test that stat calls are properly cached.
+     */
+    public function testDoctrineStatCache()
+    {
+        StreamWrapper::attachCache(new DoctrineCacheAdapter(new ArrayCache()));
+        $ts = strtotime('Tuesday, April 9 2013');
+        $this->setMockResponse($this->client, array(new Response(200, array(
+          'Content-Length' => 5,
+          'Last-Modified'  => gmdate('r', $ts)
+        ))));
+
+        // We do two stat() calls, but have only queued a single response. If the
+        // cache isn't working, the second stat() call will throw an exception.
+        clearstatcache('s3://bucket/key');
+        $stat = stat('s3://bucket/key');
+        clearstatcache('s3://bucket/key');
+        $stat = stat('s3://bucket/key');
+        $this->assertEquals(0100777, $stat['mode']);
+        $this->assertEquals(5, $stat['size']);
+        $this->assertEquals($ts, $stat['mtime']);
+        $this->assertEquals($ts, $stat['ctime']);
+        StreamWrapper::detachCache();
+    }
+
+    /**
+     * Test that cache lifetimes are respected.
+     *
+     * @expectedException \PHPUnit_Framework_Error_Warning
+     * @expectedExceptionCode 512
+     */
+    public function testStatCacheLifetime() {
+        // We can't use ArrayCache as it doesn't support lifetimes.
+        $cacheDir = sys_get_temp_dir() . '/' . uniqid('S3StreamWrapperTest-');
+        $cache = new PhpFileCache($cacheDir);
+
+        StreamWrapper::attachCache(new DoctrineCacheAdapter($cache), 1);
+        $ts = strtotime('Tuesday, April 9 2013');
+        $this->setMockResponse($this->client, array(new Response(200, array(
+          'Content-Length' => 5,
+          'Last-Modified'  => gmdate('r', $ts)
+        ))));
+
+        clearstatcache('s3://bucket/key');
+        $stat = stat('s3://bucket/key');
+        clearstatcache('s3://bucket/key');
+        sleep(2);
```

Comment:
I really don't like this. I'd expected to be able to do at least a 1 second sleep, but on my local it looks like the "1 second" cache is valid up to around 1.5 seconds later (and isn't consistent). We could implement our own cache that extends ArrayCache() and implements subsecond cache lifetimes, but that seems bit overkill. Thoughts?